src/net/socket/tcp_sock.c

```diff
@@ -374,39 +374,69 @@ static int tcp_accept(struct sock *sk, struct sockaddr *addr,
 	return 0;
 }
 
-static int tcp_write(struct tcp_sock *tcp_sk, void *buff, size_t len) {
+static int tcp_write(struct tcp_sock *tcp_sk, struct msghdr * msg) {
+	size_t iov_len;
+	int full_len;
 	void *pb;
 	struct sk_buff *skb;
 	int ret;
+	size_t skb_len;
+	int tran_len;
+	int cp_len;
+	int cp_off;
+	int i;
+
+	full_len = 0;
+	tran_len = 0;
+	for (i = 0; i < msg->msg_iovlen; i ++) {
+		full_len += msg->msg_iov[i].iov_len;
+	}
 
-	pb = buff;
-	while (len != 0) {
-		/* Previous comment: try to send wholly msg
-		 * We must pass no more than 64k bytes to underlaying IP level */
-		size_t bytes = min(len, IP_MAX_PACKET_LEN - MAX_HEADER_SIZE);
-		skb = NULL; /* alloc new pkg */
+	iov_len = 0;
+	skb_len = 0;
 
-		ret = alloc_prep_skb(tcp_sk, 0, &bytes, &skb);
-		if (ret != 0) {
-			break;
+	for (i = 0; (i < msg->msg_iovlen) || (0 != skb_len) || (0 == iov_len); ) {
```

Comment:
Is it possible to use `while (tran_len < full_len) ` here instead? I think this condition should be equal to `(i < msg->msg_iovlen) || (0 != skb_len) || (0 == iov_len)`.

Reply:
Thank a lot! You're right!